apps/agreement/contracts/disputable/Delay.sol

```diff
@@ -0,0 +1,459 @@
+        uint64 executableAt;            // Date time when the delayable can be executed
+        DelayableState state;           // Current state of the delayable
+        uint256 actionId;               // Identification number of the action in the context of the agreement
+        bytes script;                   // Action script to be executed
+    }
+
+    struct TokenBalancePermission {
+        ERC20 token;                    // ERC20 token to be used for custom permissions based on token balance
+        uint256 balance;                // Amount of tokens used for custom permissions
+    }
+
+    /**
+    * @dev Auth modifier restricting access only for addresses that can submit actions
+    */
+    modifier onlySubmitter() {
+        require(_canSubmit(msg.sender), ERROR_AUTH_FAILED);
+        _;
+    }
+
+    uint64 public delayPeriod;
+    Delayable[] private delayables;
+
+    TokenBalancePermission private submitTokenBalancePermission;
+    TokenBalancePermission private challengeTokenBalancePermission;
+
+    /**
+    * @notice Initialize Delay app with `@transformTime(_delayPeriod)`, linked to the Agreement `_agreement` with the following requirements:
+    * @notice - `@tokenAmount(_collateralToken, _actionCollateral)` collateral for submitting actions
+    * @notice - `@tokenAmount(_collateralToken, _challengeCollateral)` collateral for challenging actions
+    * @notice - `@transformTime(_challengeDuration)` for the challenge duration
+    * @notice - Submit permission: `_submitPermissionToken == 0 ? 'None' : @tokenAmount(_submitPermissionToken, _submitPermissionBalance)`
+    * @notice - Challenge per: `_challengePermissionBalance == 0 ? 'None' : @tokenAmount(_challengePermissionToken, _challengePermissionBalance)`
+    * @param _delayPeriod Duration in seconds during which an action is delayed before being executed
+    * @param _agreement Agreement instance to be linked
+    * @param _collateralToken Address of the ERC20 token to be used for collateral
+    * @param _actionCollateral Amount of collateral tokens that will be locked every time an delayable is submitted
+    * @param _challengeCollateral Amount of collateral tokens that will be locked every time an delayable is challenged
+    * @param _challengeDuration Duration in seconds of the challenge, during this time window the submitter can answer the challenge
+    * @param _submitPermissionToken ERC20 token to be used for custom submitting permissions based on token balance
+    * @param _submitPermissionBalance Amount of `_submitPermissionToken` tokens for custom signing permissions
+    * @param _challengePermissionToken ERC20 token to be used for custom challenge permissions based on token balance
+    * @param _challengePermissionBalance Amount of `_challengePermissionBalance` tokens for custom challenge permissions
+    */
+    function initialize(
+        uint64 _delayPeriod,
+        IAgreement _agreement,
```

Comment:
Something I've been thinking about is the concept of "registering" Disputables to an Agreement.

This is nice on a few levels:

- It will let us move the Agreement configuration after initialization
- It will most likely allow us to "re-register" a Disputable onto another Agreement
- We will be able to set up both forward and backward links between a Disputable and an Agreement, and with those, we will be able to "cheat" the permissions (we can use `msg.sender` and validate whether the call came from the Agreement or a registered Disputable for validation instead of going through the ACL)

Reply:
> It will most likely allow us to "re-register" a Disputable onto another Agreement

This could become a problem if there are ongoing actions. Note that the disputable app needs to tell the agreement when the action has been executed (currently "close"), e.g. the happy path for a delayable action. We could support having the agreement instance stored per action tho, but then the abstraction becomes more white box.

I agree with the other two tho :)